packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -74,15 +94,55 @@ class WebSocketConnection implements Closeable {
       Stream<dynamic> stream) {
     return stream.transform(const WebSocketMessageStreamTransformer()).listen(
       _onData,
-      cancelOnError: true,
+      onDone: () {
+        _logger.verbose('Web socket connection done.');
+
+        _resetConnectionInit();
+        _resetConnectionVariables();
+      },
       onError: (Object e) {
+        if (!_connectionReady.isCompleted) _connectionReady.completeError(e);
         _connectionError(
-          ApiException('Connection failed.', underlyingException: e.toString()),
+          ApiException('Connection failed.', underlyingException: e),
         );
+        _hubEventsController?.add(SubscriptionHubEvent.failed());
+
+        if (_hasNetwork &&
+            _hasConnectionBroken &&
+            _subscriptionRequests.isNotEmpty) {
+          _retry();
+        }
```

Comment:
Should this be:

```suggestion
        if (_hasNetwork &&
            _hasConnectionBroken &&
            _subscriptionRequests.isNotEmpty) {
          _retry();
        } else {
          _hubEventsController?.add(SubscriptionHubEvent.failed());
        }
```

Since _retry will immediately add a `connecting` event